src/tasks.cpp

```diff
@@ -48,38 +49,30 @@ void Dispatcher::threadMain()
 			taskSignal.wait(taskLockUnique);
 		}
 
-		if (!taskList.empty()) {
-			// take the first task
-			Task* task = taskList.front();
-			taskList.pop_front();
-			taskLockUnique.unlock();
+		tmpTaskList = taskList;
+		taskList.clear();
+		taskLockUnique.unlock();
```

Comment:
`std::vector<Task*> tmpTaskList = std::move(taskList);`

Reply:
taskList will be left in a unknown state(?) That also kills the optimization which both taskList and tmpTaskList keeping the reserved buffer avoiding allocations when adding tasks. Moving would prevent a single copy between vectors, but at what price?

Reply:
Hmm, that's fair, but since you are locking, what is the point of making a copy?

Reply:
To reducer lock contention. Since we got a list of tasks, it will spend n time to execute those tasks. If they are executed inside the lock, the lock will be locked for that n time delaying any other any other thread that wants the lock and even worst as executing tasks will call Dispatcher:addTask there will be a deadlock.

Reply:
It makes sense, but just declare the copy array here instead of at the top of the function. I would do something along the lines of

```
std::vector<Task*> tasksToCheck;
std::swap(taskList, tasksToCheck);
taskLockUnique.unlock();
```

Solves the unknown state and still saves the copy :P if you want to keep the reserved capacity, you can use `reserve`, but I honestly wouldn't bother.

Reply:
That probably will fix, but i don't agree. I had two aims with this PR: reduce allocation calls and lock contention.